### PR TITLE
Adding ping feature, issue fixes

### PR DESCRIPTION
**Pull request by milanbhagwat**

### CODE REVIEW

File: activity/ws/activity.go [hunk 1]
```diff
@@ -268,3 +284,44 @@ func getCerts(trustStore string) (*x509.CertPool, error) {
...
+func (a *Activity) Cleanup() error {
+	a.continuePing = false
+	a.cachedClients.Range(func(key, value interface{}) bool {
+		conn, ok := value.(*websocket.Conn)
+		if !ok {
+			logger1.Info("value is not websocket connection type to close while activity cleanup")
+		} else {
+			err := conn.WriteControl(websocket.CloseMessage, []byte("Close connection while Activity cleanup"), time.Now().Add(time.Second))
+			if err != nil {
+				logger1.Infof("error while sending close message: %v", err)
```
Comment (lixingwang) [inline activity/ws/activity.go]: Let's use warn level. this kind of error harmless, so warn would make more sense.

File: trigger/wsclient/trigger.go [hunk 2]
```diff
@@ -157,12 +181,11 @@ func (t *Trigger) Start() error {
 			}()
 			for {
 				_, message, err := t.wsconn.ReadMessage()
-				//t.logger.Infof("Message received :", string(message)) //TODO REMOVE
 				if err != nil {
 					t.logger.Errorf("error while reading websocket message: %s", err)
 					break
 				}
-
+				t.logger.Info("New message received...")
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: Just make sure if we really need this log

File: trigger/wsclient/trigger.go [hunk 3]
```diff
@@ -288,3 +313,22 @@ func decodeCerts(certVal string, log log.Logger) ([]byte, error) {
...
+func ping(connection *websocket.Conn, tr *Trigger) {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for {
+		if tr.continuePing {
+			select {
+			case t := <-ticker.C:
+				tr.logger.Debugf("Sending Ping at timestamp : %v", t)
+				if err := connection.WriteControl(websocket.PingMessage, []byte("---HeartBeat---"), time.Now().Add(time.Second)); err != nil {
+					tr.logger.Debugf("error while sending ping: %v", err)
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: Let's use error for this case.

File: activity/ws/activity.go [hunk 4]
```diff
@@ -45,15 +47,19 @@ func New(ctx activity.InitContext) (activity.Activity, error) {
...
 	cachedClients sync.Map
+	continuePing  bool
 }
 
 // Metadata returns the metadata for a websocket client
 func (a *Activity) Metadata() *activity.Metadata {
 	return activityMd
 }
 
+var logger1 log.Logger
```
Comment (lixingwang) [inline activity/ws/activity.go]: should we rename it to a better name?

File: trigger/wsclient/trigger.go [hunk 5]
```diff
@@ -134,11 +137,32 @@ func (t *Trigger) Initialize(ctx trigger.InitContext) error {
...
+		ctx.Logger().Debugf("received pong msg from server: %s", msg)
+		return nil
+	})
+	// send ping to avoid TCI connection timeout
+	go ping(conn, t)
+	/*
+		// Initiate conversation is Trigger is in Chat mode
+		if os.Getenv("CHATMODE") == "TRUE" {
+			go initiateConversation(conn, t)
+		}
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: Please remove it to avoid confuse

File: trigger/wsclient/trigger.go [hunk 6]
```diff
@@ -134,11 +137,32 @@ func (t *Trigger) Initialize(ctx trigger.InitContext) error {
...
+	*/
 	t.tInitContext = ctx
 
 	return nil
 }
 
+func initiateConversation(connection *websocket.Conn, tr *Trigger) {
```
Comment (lixingwang) [inline trigger/wsclient/trigger.go]: reemove

**Review by lixingwang — CHANGES_REQUESTED** [review]

File: trigger/wsclient/trigger.go (hunk 2 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: QA wanted to have more logs to get clarity on state while execution. make be we can make it Debug level.

File: activity/ws/activity.go (hunk 1 above)
Comment (mbhagwat-tibco) [reply activity/ws/activity.go]: Done

File: trigger/wsclient/trigger.go (hunk 3 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: done

File: trigger/wsclient/trigger.go (hunk 5 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: done

File: trigger/wsclient/trigger.go (hunk 6 above)
Comment (mbhagwat-tibco) [reply trigger/wsclient/trigger.go]: done

File: activity/ws/activity.go (hunk 4 above)
Comment (mbhagwat-tibco) [reply activity/ws/activity.go]: done

**Review by lixingwang — APPROVED** [review]

**Pull request merged**